components/dashboard/src/projects/ProjectVariables.tsx

```diff
@@ -0,0 +1,132 @@
+                <button onClick={() => setShowAddVariableModal(true)}>New Variable</button>
+            </div>
+            : <>
+                <ItemsList>
+                    <Item header={true} className="grid grid-cols-4 items-center">
+                        <ItemField>Name</ItemField>
+                        <ItemField>Value</ItemField>
+                        <ItemField>Visible in Workspaces?</ItemField>
+                        <ItemField></ItemField>
+                    </Item>
+                    {envVars.map(variable => {
+                        return <Item className="grid grid-cols-4 items-center">
+                            <ItemField>{variable.name}</ItemField>
+                            <ItemField>****</ItemField>
+                            <ItemField>{variable.censored ? 'Hidden' : 'Visible'}</ItemField>
+                            <ItemField className="flex justify-end">
+                                <ItemFieldContextMenu menuEntries={[
+                                    {
+                                        title: 'Delete',
+                                        customFontStyle: 'text-red-600 dark:text-red-400 hover:text-red-800 dark:hover:text-red-300',
+                                        onClick: () => deleteEnvVar(variable.id),
+                                    },
+                                ]} />
+                            </ItemField>
+                        </Item>
+                    })}
+                </ItemsList>
+            </>
+        }
+    </ProjectSettingsPage>;
+}
+
+function AddVariableModal(props: { project?: Project, onClose: () => void }) {
+    const [ name, setName ] = useState<string>("");
+    const [ value, setValue ] = useState<string>("");
+    const [ censored, setCensored ] = useState<boolean>(true);
+    const [ error, setError ] = useState<Error | undefined>();
+
+    const addVariable = async () => {
+        if (!props.project) {
+            return;
+        }
+        try {
+            await getGitpodService().server.setProjectEnvironmentVariable(props.project.id, name, value, censored);
+            props.onClose();
+        } catch (err) {
+            setError(err);
+        }
+    }
+
+    return <Modal visible={true} onClose={props.onClose} onEnter={() => { addVariable(); return false; }}>
+        <h3 className="mb-4">Add Variable</h3>
+        <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 py-4 flex flex-col">
+            <AlertBox><strong>Project variables might be accessible by anyone with read access to your repository.</strong><br/>Secret values can be exposed if they are printed in the logs, persisted to the file system, or made visible in workspaces.</AlertBox>
```

Comment:
**thought**: It took me a while to wrap my head around the security cases here. 😇 

Some improvements could include:
1. Making the alert title brief and short in one line.
2. Avoiding complex state handling for users by default.

**suggestion**: For the second point I still think that having a sinsible default where variables can be exposed everywhere instead of requiring users to think everytime could be better. Here's how the copy could look like although I could be wrong on terminal access only:

| BEFORE | AFTER |
|-|-|
| <img width="1440" alt="Screenshot 2022-01-14 at 8 01 26 PM (2)" src="https://user-images.githubusercontent.com/120486/149563824-b64f2853-19ad-479b-bfe0-cb060c5bfa71.png"> | <img width="1440" alt="Screenshot 2022-01-14 at 8 01 10 PM (2)" src="https://user-images.githubusercontent.com/120486/149563836-8b476640-7ff5-447d-a9a4-6d0c54bc1af2.png"> |

Let me know what you think!

Reply:
- The message at the top could be a little clearer about how prebuild variables could leak into workspaces.
- The current default to **Hide in workspaces** is good IMO because it errs on the side of less exposure.
- I think we can leave out the extra instruction for what it means to uncheck the checkbox.

Suggest the following:

> **Warning**  
> Even with **Hide in Workspaces** checked, a project-specific environment variable may become exposed to anyone who can start a workspace e.g. if logged or persisted to the file system. 

> Name
> [ _ _ _ ]

> Value
> [ _ _ _ ]

> ☑️ **Hide in Workspaces**
> _Project-specific environment variables are always visible in prebuilds._

(after unchecking)

> **NOTE**
> This environment variable will be visible to anyone who starts a Gitpod workspace on this repository.

Reply:
Thanks for the excellent feedback here! 🙏

_Phew._ After splitting a few more hairs on this, I think I've reached an optimum consensus solution on the wording:

<table><tr><td>
<img width="1435" alt="Screenshot 2022-01-17 at 19 09 11" src="https://user-images.githubusercontent.com/599268/149819846-18fb1104-3b3d-44e5-90ef-4f7a1d57c3bc.png">
</td><td>
<img width="1437" alt="Screenshot 2022-01-17 at 19 09 19" src="https://user-images.githubusercontent.com/599268/149819851-8fc39ca6-762f-4de5-8dd0-5447e6510ac7.png">
</td></tr></table>

Going to ship this 🚢 (but happy to further split hairs in follow-up PRs 😅)